### PR TITLE
feat: update redirects

### PR DESCRIPTION
**Pull request by chicharr**
Feature to update redirects spreadsheet for an Edge Delivery Services website.
Receives a list of new redirects, validates them, ensuring they have the right format, and that they will not duplicate or create cycles with the existing redirect rules, and appends them to the existing rules in the spreadsheet.

### CODE REVIEW

File: packages/spacecat-shared-content-client/src/clients/content-client.js [hunk 1]
```diff
@@ -97,6 +98,68 @@ const validateMetadata = (metadata) => {
...
+    .forEach((redirectRule) => redirectsSet.add(redirectRule));
+
+  const newRedirectsClean = [];
+  newRedirects.forEach((redirectRule) => {
+    const { from, to } = redirectRule;
+    const strRedirectRule = `${from}:${to}`;
+    if (!redirectsSet.has(strRedirectRule)) {
+      redirectsSet.add(strRedirectRule);
+      newRedirectsClean.push(redirectRule);
+    } // TODO: report duplicated entry??
```
Comment (alinarublea) [inline packages/spacecat-shared-content-client/src/clients/content-client.js:149]: Here I would log the duplicate entry

File: packages/spacecat-shared-content-client/src/clients/content-client.js [hunk 2]
```diff
@@ -97,6 +98,68 @@ const validateMetadata = (metadata) => {
...
+  });
+  return newRedirectsClean;
+};
+
+const detectRedirectLoops = (currentRedirects, newRedirects) => {
+  const redirectsGraph = new Graph();
+  [...currentRedirects, ...newRedirects].forEach((r) => redirectsGraph.addEdge(r.from, r.to));
+  if (hasCycle(redirectsGraph)) {
+    // TODO: report cycle path??
+    throw new Error('Redirect cycle detected');
```
Comment (alinarublea) [inline packages/spacecat-shared-content-client/src/clients/content-client.js:159]: In this case I see we are not adding any of the entries. Could we check after each edge if there is a cycle and just not add that specific edge ?

**Review by alinarublea — APPROVED** [review]
Except the small comments all the rest looks good to me

File: packages/spacecat-shared-content-client/src/clients/content-client.js [hunk 3]
```diff
@@ -97,6 +98,80 @@ const validateMetadata = (metadata) => {
...
+
+    if (redirect.from === redirect.to) {
+      throw new Error('Redirect from and to paths must be different');
+    }
+  }
+};
+
+const removeDuplicatedRedirects = (currentRedirects, newRedirects, log) => {
+  const redirectsSet = new Set();
+  currentRedirects.map(({ from, to }) => `${from}:${to}`)
```
Comment (dipratap) [inline packages/spacecat-shared-content-client/src/clients/content-client.js:139]: Can be simplified to
```
const redirectsSet = new Set(
    currentRedirects.map(({ from, to }) => `${from}:${to}`)
);
```

File: packages/spacecat-shared-content-client/src/clients/content-client.js [hunk 4]
```diff
@@ -97,6 +98,80 @@ const validateMetadata = (metadata) => {
...
+
+    if (!hasText(redirect.from)) {
+      throw new Error('Redirect must have a valid from path');
+    }
+
+    if (!hasText(redirect.to)) {
+      throw new Error('Redirect must have a valid to path');
+    }
+
+    if (!redirect.from.startsWith('/')) {
```
Comment (dipratap) [inline packages/spacecat-shared-content-client/src/clients/content-client.js:123]: Can we use a regex to validate the from and to endpoints?

File: packages/spacecat-shared-content-client/src/clients/content-client.js [hunk 5]
```diff
@@ -192,4 +269,50 @@ export default class ContentClient {
...
+    if (cleanNewRedirects.length === 0) {
+      this.log.info('No valid redirects to update');
+      return;
+    }
+    const noCycleRedirects = removeRedirectLoops(currentRedirects, cleanNewRedirects, this.log);
+    if (noCycleRedirects.length === 0) {
+      this.log.info('No valid redirects to update');
+      return;
+    }
+    if (noCycleRedirects.length !== cleanNewRedirects.length) {
```
Comment (dipratap) [inline packages/spacecat-shared-content-client/src/clients/content-client.js:307]: Should we log the removed loops in the removeRedirectLoops method itself?

**Review by dipratap — APPROVED** [review]

**Pull request merged**